File: mobile/androidApp/app/src/main/java/com/pipertool/viewmodel/SettingsViewModel.kt
package com.pipertool.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.util.Date
import javax.inject.Inject

sealed class SettingsState {
    object Idle : SettingsState()
    object Loading : SettingsState()
    object Success : SettingsState()
    data class Error(val message: String) : SettingsState()
}

data class AccountSettings(
    val id: String,
    val name: String,
    val email: String,
    val avatarUrl: String?,
    val isPremium: Boolean,
    val createdAt: Date,
    val lastLoginAt: Date
)

data class AppearanceSettings(
    val theme: String = "system", // "system", "light", "dark"
    val language: String = "English",
    val useSystemFontSize: Boolean = true,
    val reduceMotion: Boolean = false,
    val defaultView: String = "list" // "list", "grid", "compact"
)

data class NotificationSettings(
    val emailNotifications: Boolean = true,
    val pushNotifications: Boolean = true,
    val newsletterReminders: Boolean = true,
    val newContentAlerts: Boolean = true,
    val quietHoursEnabled: Boolean = false,
    val quietHoursStart: String? = null,
    val quietHoursEnd: String? = null
)

data class DataManagementSettings(
    val autoBackup: Boolean = true,
    val backupFrequency: String = "weekly", // "daily", "weekly", "monthly"
    val lastBackupAt: Date? = null,
    val cacheSize: Long = 0 // in bytes
)

data class PrivacySettings(
    val twoFactorEnabled: Boolean = false,
    val analyticsEnabled: Boolean = true,
    val crashReportingEnabled: Boolean = true,
    val shareUsageData: Boolean = false
)

data class AdvancedSettings(
    val developerMode: Boolean = false,
    val debugLogging: Boolean = false,
    val experimentalFeatures: Boolean = false,
    val autoUpdate: Boolean = true
)

data class AllSettings(
    val account: AccountSettings?,
    val appearance: AppearanceSettings,
    val notifications: NotificationSettings,
    val dataManagement: DataManagementSettings,
    val privacy: PrivacySettings,
    val advanced: AdvancedSettings
)

@HiltViewModel
class SettingsViewModel @Inject constructor(
    private val settingsManager: SettingsManager
) : ViewModel() {
    
    private val _state = MutableStateFlow<SettingsState>(SettingsState.Idle)
    val state: StateFlow<SettingsState> = _state.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _account = MutableStateFlow<AccountSettings?>(null)
    val account: StateFlow<AccountSettings?> = _account.asStateFlow()
    
    private val _appearance = MutableStateFlow(AppearanceSettings())
    val appearance: StateFlow<AppearanceSettings> = _appearance.asStateFlow()
    
    private val _notifications = MutableStateFlow(NotificationSettings())
    val notifications: StateFlow<NotificationSettings> = _notifications.asStateFlow()
    
    private val _dataManagement = MutableStateFlow(DataManagementSettings())
    val dataManagement: StateFlow<DataManagementSettings> = _dataManagement.asStateFlow()
    
    private val _privacy = MutableStateFlow(PrivacySettings())
    val privacy: StateFlow<PrivacySettings> = _privacy.asStateFlow()
    
    private val _advanced = MutableStateFlow(AdvancedSettings())
    val advanced: StateFlow<AdvancedSettings> = _advanced.asStateFlow()
    
    init {
        setupAutoSave()
        loadSettings()
    }
    
    private fun setupAutoSave() {
        // Auto-save appearance changes
        viewModelScope.launch {
            _appearance
                .drop(1) // Skip initial value
                .debounce(1000) // Wait 1 second after last change
                .collect { saveSettings() }
        }
        
        // Auto-save notification changes
        viewModelScope.launch {
            _notifications
                .drop(1)
                .debounce(1000)
                .collect { saveSettings() }
        }
        
        // Auto-save privacy changes
        viewModelScope.launch {
            _privacy
                .drop(1)
                .debounce(1000)
                .collect { saveSettings() }
        }
        
        // Auto-save advanced changes
        viewModelScope.launch {
            _advanced
                .drop(1)
                .debounce(1000)
                .collect { saveSettings() }
        }
    }
    
    fun loadSettings() {
        viewModelScope.launch {
            _isLoading.value = true
            _state.value = SettingsState.Loading
            
            try {
                val settings = settingsManager.fetchSettings()
                updateSettings(settings)
                _state.value = SettingsState.Success
            } catch (e: Exception) {
                _state.value = SettingsState.Error(e.message ?: "Failed to load settings")
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    private fun updateSettings(settings: AllSettings) {
        _account.value = settings.account
        _appearance.value = settings.appearance
        _notifications.value = settings.notifications
        _dataManagement.value = settings.dataManagement
        _privacy.value = settings.privacy
        _advanced.value = settings.advanced
    }
    
    private fun saveSettings() {
        viewModelScope.launch {
            try {
                val allSettings = AllSettings(
                    account = _account.value,
                    appearance = _appearance.value,
                    notifications = _notifications.value,
                    dataManagement = _dataManagement.value,
                    privacy = _privacy.value,
                    advanced = _advanced.value
                )
                
                settingsManager.saveSettings(allSettings)
            } catch (e: Exception) {
                // Handle save error silently or show user notification
                println("Failed to save settings: ${e.message}")
            }
        }
    }
    
    // Appearance update methods
    fun updateTheme(theme: String) {
        _appearance.value = _appearance.value.copy(theme = theme)
    }
    
    fun updateLanguage(language: String) {
        _appearance.value = _appearance.value.copy(language = language)
    }
    
    fun updateUseSystemFontSize(useSystem: Boolean) {
        _appearance.value = _appearance.value.copy(useSystemFontSize = useSystem)
    }
    
    fun updateReduceMotion(reduce: Boolean) {
        _appearance.value = _appearance.value.copy(reduceMotion = reduce)
    }
    
    // Notification update methods
    fun updateEmailNotifications(enabled: Boolean) {
        _notifications.value = _notifications.value.copy(emailNotifications = enabled)
    }
    
    fun updatePushNotifications(enabled: Boolean) {
        _notifications.value = _notifications.value.copy(pushNotifications = enabled)
    }
    
    fun updateNewsletterReminders(enabled: Boolean) {
        _notifications.value = _notifications.value.copy(newsletterReminders = enabled)
    }
    
    fun updateNewContentAlerts(enabled: Boolean) {
        _notifications.value = _notifications.value.copy(newContentAlerts = enabled)
    }
    
    // Data management update methods
    fun updateAutoBackup(enabled: Boolean) {
        _dataManagement.value = _dataManagement.value.copy(autoBackup = enabled)
    }
    
    fun updateCacheSize(size: Long) {
        _dataManagement.value = _dataManagement.value.copy(cacheSize = size)
    }
    
    // Privacy update methods
    fun updateTwoFactor(enabled: Boolean) {
        _privacy.value = _privacy.value.copy(twoFactorEnabled = enabled)
    }
    
    fun updateAnalytics(enabled: Boolean) {
        _privacy.value = _privacy.value.copy(analyticsEnabled = enabled)
    }
    
    fun updateCrashReporting(enabled: Boolean) {
        _privacy.value = _privacy.value.copy(crashReportingEnabled = enabled)
    }
    
    fun updateShareUsageData(enabled: Boolean) {
        _privacy.value = _privacy.value.copy(shareUsageData = enabled)
    }
    
    // Advanced update methods
    fun updateDeveloperMode(enabled: Boolean) {
        _advanced.value = _advanced.value.copy(developerMode = enabled)
    }
    
    fun updateDebugLogging(enabled: Boolean) {
        _advanced.value = _advanced.value.copy(debugLogging = enabled)
    }
    
    fun updateExperimentalFeatures(enabled: Boolean) {
        _advanced.value = _advanced.value.copy(experimentalFeatures = enabled)
    }
    
    fun updateAutoUpdate(enabled: Boolean) {
        _advanced.value = _advanced.value.copy(autoUpdate = enabled)
    }
    
    // Data operations
    fun exportData(format: String) {
        viewModelScope.launch {
            try {
                settingsManager.exportData(format)
                // Handle success - could show a notification
            } catch (e: Exception) {
                _state.value = SettingsState.Error("Failed to export data: ${e.message}")
            }
        }
    }
    
    fun clearCache() {
        viewModelScope.launch {
            try {
                settingsManager.clearCache()
                _dataManagement.value = _dataManagement.value.copy(cacheSize = 0)
            } catch (e: Exception) {
                _state.value = SettingsState.Error("Failed to clear cache: ${e.message}")
            }
        }
    }
    
    fun resetSettings() {
        viewModelScope.launch {
            try {
                settingsManager.resetSettings()
                loadSettings() // Reload default settings
            } catch (e: Exception) {
                _state.value = SettingsState.Error("Failed to reset settings: ${e.message}")
            }
        }
    }
    
    fun syncSettings() {
        viewModelScope.launch {
            try {
                settingsManager.syncSettings()
            } catch (e: Exception) {
                _state.value = SettingsState.Error("Failed to sync settings: ${e.message}")
            }
        }
    }
    
    fun signOut() {
        viewModelScope.launch {
            try {
                settingsManager.signOut()
                // Navigate to auth screen would typically be handled by the UI
            } catch (e: Exception) {
                _state.value = SettingsState.Error("Failed to sign out: ${e.message}")
            }
        }
    }
}

class SettingsManager @Inject constructor() {
    private val mockDelay = 500L
    
    suspend fun fetchSettings(): AllSettings {
        kotlinx.coroutines.delay(mockDelay)
        
        val account = AccountSettings(
            id = "user_123",
            name = "John Doe",
            email = "john.doe@example.com",
            avatarUrl = "https://example.com/avatar.jpg",
            isPremium = true,
            createdAt = Date(System.currentTimeMillis() - 86400000L * 30), // 30 days ago
            lastLoginAt = Date()
        )
        
        val appearance = AppearanceSettings()
        
        val notifications = NotificationSettings()
        
        val dataManagement = DataManagementSettings(
            autoBackup = true,
            backupFrequency = "weekly",
            lastBackupAt = Date(System.currentTimeMillis() - 86400000L * 3), // 3 days ago
            cacheSize = 1024L * 1024L * 50 // 50MB
        )
        
        val privacy = PrivacySettings()
        
        val advanced = AdvancedSettings()
        
        return AllSettings(
            account = account,
            appearance = appearance,
            notifications = notifications,
            dataManagement = dataManagement,
            privacy = privacy,
            advanced = advanced
        )
    }
    
    suspend fun saveSettings(settings: AllSettings) {
        kotlinx.coroutines.delay(mockDelay)
        // Mock implementation - in real app, save to server/local storage
    }
    
    suspend fun exportData(format: String) {
        kotlinx.coroutines.delay(mockDelay)
        // Mock implementation - generate export file
    }
    
    suspend fun importData(filePath: String) {
        kotlinx.coroutines.delay(mockDelay)
        // Mock implementation - parse and import data
    }
    
    suspend fun clearCache() {
        kotlinx.coroutines.delay(mockDelay)
        // Mock implementation - clear app cache
    }
    
    suspend fun syncSettings() {
        kotlinx.coroutines.delay(mockDelay)
        // Mock implementation - sync with server
    }
    
    suspend fun resetSettings() {
        kotlinx.coroutines.delay(mockDelay)
        // Mock implementation - reset to defaults
    }
    
    suspend fun signOut() {
        kotlinx.coroutines.delay(mockDelay)
        // Mock implementation - clear user data
    }
}

File: mobile/iosApp/iosApp/ViewModels/SettingsViewModel.swift
import SwiftUI
import Combine

enum SettingsState {
    case idle
    case loading
    case success
    case error(String)
}

struct AccountSettings: Codable {
    let id: String
    let name: String
    let email: String
    let avatarUrl: String?
    let isPremium: Bool
    let createdAt: Date
    let lastLoginAt: Date
}

struct AppearanceSettings: Codable {
    var theme: String // "system", "light", "dark"
    var language: String
    var useSystemFontSize: Bool
    var reduceMotion: Bool
    var defaultView: String // "list", "grid", "compact"
}

struct NotificationSettings: Codable {
    var emailNotifications: Bool
    var pushNotifications: Bool
    var newsletterReminders: Bool
    var newContentAlerts: Bool
    var quietHoursEnabled: Bool
    var quietHoursStart: String?
    var quietHoursEnd: String?
}

struct DataManagementSettings: Codable {
    var autoBackup: Bool
    var backupFrequency: String // "daily", "weekly", "monthly"
    var lastBackupAt: Date?
    var cacheSize: Int64 // in bytes
}

struct PrivacySettings: Codable {
    var twoFactorEnabled: Bool
    var analyticsEnabled: Bool
    var crashReportingEnabled: Bool
    var shareUsageData: Bool
}

struct AdvancedSettings: Codable {
    var developerMode: Bool
    var debugLogging: Bool
    var experimentalFeatures: Bool
    var autoUpdate: Bool
}

class SettingsViewModel: ObservableObject {
    @Published private(set) var state: SettingsState = .idle
    @Published private(set) var isLoading: Bool = false
    @Published private(set) var account: AccountSettings?
    
    @Published var appearance = AppearanceSettings(
        theme: "system",
        language: "English",
        useSystemFontSize: true,
        reduceMotion: false,
        defaultView: "list"
    )
    
    @Published var notifications = NotificationSettings(
        emailNotifications: true,
        pushNotifications: true,
        newsletterReminders: true,
        newContentAlerts: true,
        quietHoursEnabled: false,
        quietHoursStart: nil,
        quietHoursEnd: nil
    )
    
    @Published var dataManagement = DataManagementSettings(
        autoBackup: true,
        backupFrequency: "weekly",
        lastBackupAt: nil,
        cacheSize: 0
    )
    
    @Published var privacy = PrivacySettings(
        twoFactorEnabled: false,
        analyticsEnabled: true,
        crashReportingEnabled: true,
        shareUsageData: false
    )
    
    @Published var advanced = AdvancedSettings(
        developerMode: false,
        debugLogging: false,
        experimentalFeatures: false,
        autoUpdate: true
    )
    
    private let settingsManager = SettingsManager()
    private var cancellables = Set<AnyCancellable>()
    
    init() {
        setupBindings()
    }
    
    private func setupBindings() {
        // Auto-save appearance changes
        $appearance
            .dropFirst()
            .debounce(for: .seconds(1), scheduler: RunLoop.main)
            .sink { [weak self] _ in
                self?.saveSettings()
            }
            .store(in: &cancellables)
        
        // Auto-save notification changes
        $notifications
            .dropFirst()
            .debounce(for: .seconds(1), scheduler: RunLoop.main)
            .sink { [weak self] _ in
                self?.saveSettings()
            }
            .store(in: &cancellables)
        
        // Auto-save privacy changes
        $privacy
            .dropFirst()
            .debounce(for: .seconds(1), scheduler: RunLoop.main)
            .sink { [weak self] _ in
                self?.saveSettings()
            }
            .store(in: &cancellables)
        
        // Auto-save advanced changes
        $advanced
            .dropFirst()
            .debounce(for: .seconds(1), scheduler: RunLoop.main)
            .sink { [weak self] _ in
                self?.saveSettings()
            }
            .store(in: &cancellables)
    }
    
    func loadSettings() {
        isLoading = true
        state = .loading
        
        settingsManager.fetchSettings()
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    self?.isLoading = false
                    if case .failure(let error) = completion {
                        self?.state = .error(error.localizedDescription)
                    }
                },
                receiveValue: { [weak self] settings in
                    self?.updateSettings(settings)
                    self?.state = .success
                }
            )
            .store(in: &cancellables)
    }
    
    private func updateSettings(_ settings: AllSettings) {
        account = settings.account
        appearance = settings.appearance
        notifications = settings.notifications
        dataManagement = settings.dataManagement
        privacy = settings.privacy
        advanced = settings.advanced
    }
    
    func saveSettings() {
        let allSettings = AllSettings(
            account: account,
            appearance: appearance,
            notifications: notifications,
            dataManagement: dataManagement,
            privacy: privacy,
            advanced: advanced
        )
        
        settingsManager.saveSettings(allSettings)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { completion in
                    if case .failure(let error) = completion {
                        print("Failed to save settings: \(error)")
                    }
                },
                receiveValue: { _ in }
            )
            .store(in: &cancellables)
    }
    
    func exportData(format: String) {
        settingsManager.exportData(format: format)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { completion in
                    if case .failure(let error) = completion {
                        print("Failed to export data: \(error)")
                    }
                },
                receiveValue: { url in
                    // Handle exported file URL
                    print("Data exported to: \(url)")
                }
            )
            .store(in: &cancellables)
    }
    
    func importData(from url: URL) {
        settingsManager.importData(from: url)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    if case .failure(let error) = completion {
                        self?.state = .error(error.localizedDescription)
                    } else {
                        self?.loadSettings() // Reload settings after import
                    }
                },
                receiveValue: { _ in }
            )
            .store(in: &cancellables)
    }
    
    func clearCache() {
        settingsManager.clearCache()
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { completion in
                    if case .failure(let error) = completion {
                        print("Failed to clear cache: \(error)")
                    } else {
                        // Update cache size
                        dataManagement.cacheSize = 0
                    }
                },
                receiveValue: { _ in }
            )
            .store(in: &cancellables)
    }
    
    func syncSettings() {
        settingsManager.syncSettings()
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { completion in
                    if case .failure(let error) = completion {
                        print("Failed to sync settings: \(error)")
                    }
                },
                receiveValue: { _ in }
            )
            .store(in: &cancellables)
    }
    
    func resetSettings() {
        settingsManager.resetSettings()
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    if case .failure(let error) = completion {
                        self?.state = .error(error.localizedDescription)
                    } else {
                        self?.loadSettings() // Reload default settings
                    }
                },
                receiveValue: { _ in }
            )
            .store(in: &cancellables)
    }
    
    func signOut() {
        settingsManager.signOut()
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { _ in },
                receiveValue: { _ in }
            )
            .store(in: &cancellables)
    }
}

struct AllSettings: Codable {
    let account: AccountSettings?
    let appearance: AppearanceSettings
    let notifications: NotificationSettings
    let dataManagement: DataManagementSettings
    let privacy: PrivacySettings
    let advanced: AdvancedSettings
}

class SettingsManager {
    private let mockDelay: TimeInterval = 0.5
    private let userDefaults = UserDefaults.standard
    
    func fetchSettings() -> AnyPublisher<AllSettings, Error> {
        return Future<AllSettings, Error> { promise in
            DispatchQueue.global().asyncAfter(deadline: .now() + self.mockDelay) {
                let account = AccountSettings(
                    id: "user_123",
                    name: "John Doe",
                    email: "john.doe@example.com",
                    avatarUrl: "https://example.com/avatar.jpg",
                    isPremium: true,
                    createdAt: Date().addingTimeInterval(-86400 * 30),
                    lastLoginAt: Date()
                )
                
                let appearance = AppearanceSettings(
                    theme: "system",
                    language: "English",
                    useSystemFontSize: true,
                    reduceMotion: false,
                    defaultView: "list"
                )
                
                let notifications = NotificationSettings(
                    emailNotifications: true,
                    pushNotifications: true,
                    newsletterReminders: true,
                    newContentAlerts: true,
                    quietHoursEnabled: false,
                    quietHoursStart: nil,
                    quietHoursEnd: nil
                )
                
                let dataManagement = DataManagementSettings(
                    autoBackup: true,
                    backupFrequency: "weekly",
                    lastBackupAt: Date().addingTimeInterval(-86400 * 3),
                    cacheSize: 1024 * 1024 * 50 // 50MB
                )
                
                let privacy = PrivacySettings(
                    twoFactorEnabled: false,
                    analyticsEnabled: true,
                    crashReportingEnabled: true,
                    shareUsageData: false
                )
                
                let advanced = AdvancedSettings(
                    developerMode: false,
                    debugLogging: false,
                    experimentalFeatures: false,
                    autoUpdate: true
                )
                
                let allSettings = AllSettings(
                    account: account,
                    appearance: appearance,
                    notifications: notifications,
                    dataManagement: dataManagement,
                    privacy: privacy,
                    advanced: advanced
                )
                
                promise(.success(allSettings))
            }
        }
        .eraseToAnyPublisher()
    }
    
    func saveSettings(_ settings: AllSettings) -> AnyPublisher<Void, Error> {
        return Future<Void, Error> { promise in
            DispatchQueue.global().asyncAfter(deadline: .now() + self.mockDelay) {
                // Mock implementation - in real app, save to server/local storage
                promise(.success(()))
            }
        }
        .eraseToAnyPublisher()
    }
    
    func exportData(format: String) -> AnyPublisher<URL, Error> {
        return Future<URL, Error> { promise in
            DispatchQueue.global().asyncAfter(deadline: .now() + self.mockDelay) {
                let tempURL = FileManager.default.temporaryDirectory
                    .appendingPathComponent("newsletter_data")
                    .appendingPathExtension(format)
                
                // Mock data export
                let mockData = "Mock export data in \(format) format"
                try? mockData.write(to: tempURL, atomically: true, encoding: .utf8)
                
                promise(.success(tempURL))
            }
        }
        .eraseToAnyPublisher()
    }
    
    func importData(from url: URL) -> AnyPublisher<Void, Error> {
        return Future<Void, Error> { promise in
            DispatchQueue.global().asyncAfter(deadline: .now() + self.mockDelay) {
                // Mock implementation - in real app, parse and import data
                promise(.success(()))
            }
        }
        .eraseToAnyPublisher()
    }
    
    func clearCache() -> AnyPublisher<Void, Error> {
        return Future<Void, Error> { promise in
            DispatchQueue.global().asyncAfter(deadline: .now() + self.mockDelay) {
                // Mock implementation - clear app cache
                promise(.success(()))
            }
        }
        .eraseToAnyPublisher()
    }
    
    func syncSettings() -> AnyPublisher<Void, Error> {
        return Future<Void, Error> { promise in
            DispatchQueue.global().asyncAfter(deadline: .now() + self.mockDelay) {
                // Mock implementation - sync with server
                promise(.success(()))
            }
        }
        .eraseToAnyPublisher()
    }
    
    func resetSettings() -> AnyPublisher<Void, Error> {
        return Future<Void, Error> { promise in
            DispatchQueue.global().asyncAfter(deadline: .now() + self.mockDelay) {
                // Mock implementation - reset to defaults
                promise(.success(()))
            }
        }
        .eraseToAnyPublisher()
    }
    
    func signOut() -> AnyPublisher<Void, Error> {
        return Future<Void, Error> { promise in
            DispatchQueue.global().asyncAfter(deadline: .now() + self.mockDelay) {
                // Mock implementation - clear user data
                promise(.success(()))
            }
        }
        .eraseToAnyPublisher()
    }
}